### PR TITLE
Module 7 improvements

### PR DESCRIPTION
**Pull request by jack89roberts**
- Add separate section on linters
- Replace Hungarian notation with short discussion of type annotations
- Improve bad boids exercise instructions (and code)
- Remove the need for wildcard imports in some notebooks
- Make patterns an "advanced"/optional topic

Does most of #90

### CODE REVIEW

File: module07_construction_and_design/07_01_comments.ipynb [hunk 1]
```diff
@@ -7,22 +7,6 @@
     "# Comments"
```
Comment (jemrobinson) [inline module07_construction_and_design/07_01_comments.ipynb:7]: <p>I normally write comments that line-up with the start of the following line. What do you think?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/rse-course/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

File: module07_construction_and_design/07_05_object_oriented_design.ipynb [hunk 2]
```diff
@@ -7,22 +7,6 @@
     "# Design"
```
Comment (jemrobinson) [inline module07_construction_and_design/07_05_object_oriented_design.ipynb:7]: <p>This comment is very important! Worth emphasising with bold or warning signs?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/rse-course/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: module07_construction_and_design/07_07_design_patterns.ipynb [hunk 3]
```diff
@@ -4,7 +4,9 @@
    "cell_type": "markdown",
```
Comment (jemrobinson) [inline module07_construction_and_design/07_07_design_patterns.ipynb:4]: <p>There's a good webpage about design patterns here: https://refactoring.guru/design-patterns</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/alan-turing-institute/rse-course/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Review by jemrobinson — APPROVED** [review]
Again, this all looks good to me. A couple of minor comments/questions but the content all looks great!

File: module07_construction_and_design/07_01_comments.ipynb (hunk 1 above)
Comment (jack89roberts) [reply module07_construction_and_design/07_01_comments.ipynb:7]: agreed and fixed

File: module07_construction_and_design/07_07_design_patterns.ipynb (hunk 3 above)
Comment (jack89roberts) [reply module07_construction_and_design/07_07_design_patterns.ipynb:4]: <p>added</p>

**Pull request merged**